### PR TITLE
#162685647 Unit Tests for all Functionality

### PR DESCRIPTION
**Pull request by fahadmak**
#### What does this PR do?
This PR delivers a chore that writes unit tests for ALL functionality in the codebase

#### Description of Task to be completed?
Write  tests for ALL functionality that will be implemented in the codebase

#### How should this be manually tested?
Follow these steps in the command terminal:-

1. Clone this repository by running in this command below
`git clone https://github.com/andela/Ah-backend-guardians.git`

2. Change working directory to Ah-backend-guardians
`cd Ah-backend-guardians`

3. Create a virtual environment.
 `virtualenv env`

4. Activate the virtual environment
` . env/bin/activate`

5. Install the dependencies in the requirements.txt file using pip
`pip install -r requirements.txt`

6. Run this command
`python manage.py test --with-coverage`

#### What are the relevant pivotal tracker stories?
[#162685647](https://www.pivotaltracker.com/story/show/162685647)

#### Screenshots (if appropriate)
screen shot 2019-01-09 at 3 24 10 pm

### CODE REVIEW

File: authors/apps/authentication/tests/test_models.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        """ Creates an instance of the User """
+
+        self.user = User(username="test_user", email="test.sims@andela.com", password="testman101")
+        self.user.save()
+    
+
+    def test_user_creation(self):
+        """test that user is created"""
+
+        self.assertTrue(self.user, User)
```
Comment (BarnaTB) [inline authors/apps/authentication/tests/test_models.py]: Great work @fahadmak but I think this test is redundant because it is not testing any output but rather testing what you've already done in the `setUp` method. It'd be great if you removed it.

File: authors/apps/authentication/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
+from rest_framework.test import APIClient, APITestCase
+from rest_framework.views import status
+
+from ..models import User
+from ..serializers import RegistrationSerializer
+
+
+class BaseViewTest(APITestCase):
```
Comment (BarnaTB) [inline authors/apps/authentication/tests/test_views.py]: I think this class, along with all it's contents, should be in its own file probably called `base.py` since it is not testing anything but rather returning responses which are going to be tested by other classes.

File: authors/apps/authentication/tests/test_models.py (hunk 1 above)
Comment (fahadmak) [reply authors/apps/authentication/tests/test_models.py]: Thanks for the observation let me look into that

**Review by BarnaTB — CHANGES_REQUESTED** [review]
This is great work though I think you have too many classes for very minute components. You could combine registration and login into `Authentication` so that you test it all as one component and simply have methods that test each particular part that you wish to test.

Great work though!

**Review by tomuhenry — COMMENTED** [review]
@fahadmak Work on what @BarnaTB pointed out

**Review by BarnaTB — APPROVED** [review]
LGTM! Great work Fahad.

**Review by myrdstom — APPROVED** [review]

**Review by myrdstom — COMMENTED** [review]
Fahad, shouldn't we be setting up test coverage locally?

**Review by myrdstom — COMMENTED** [review]
Fahad, try getting coverage to 90%+

**Comment by VivianDoreen** [comment]
LGTM

File: authors/apps/authentication/tests/test_base.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+from rest_framework.test import APIClient, APITestCase
+
+from ..models import User
+from ..serializers import RegistrationSerializer
+
+
+class BaseViewTest(APITestCase):
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_base.py]: @fahadmak This cannot be called `BaseViewTest` please change it to
```suggestion
class BaseTestCase(APITestCase):
```
And also move this to a` base.py` file instead of test_base.py.

**Review by BarnaTB — APPROVED** [review]
LGTM! Great work @fahadmak !

File: authors/apps/authentication/tests/test_base.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    
+    client = APIClient()
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_base.py]: Please remove this irrelevant line as there is a client attribute on the `APITestCase` class available already.

File: authors/apps/authentication/tests/test_base.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                "email": email, 
+                "password": password
+                }
+            }
+        response = self.client.post(url, user_credentials, format="json")
+
+        return response
+
+
+    def setUp(self):
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_base.py]: This should be at the top of the class since it is the constructor.

File: authors/apps/authentication/tests/test_base.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+        # create user instance
+        self.user = User.objects.create(
+            username="jealous", 
+            email="simulationtys.sims@andela.com", 
+            password="simulman2101"
+            )
+        
+    
+
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_base.py]: Always had a new line at the end of the file. You can configure your editor to do that for you.

File: authors/apps/authentication/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+from django.test import TestCase
+
+from ..models import User
+
+class UserModelTest(TestCase):
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_models.py]: Always end the test classes with the word `TestCase`
```suggestion
class UserModelTestCase(TestCase):
```

File: authors/apps/authentication/tests/test_models.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    
+
+    def test_user_creation(self):
+        """test that user is created"""
+        self.assertTrue(self.user.username, "test_user")
+
+    def test_user_respresentations(self):
+        """test that user object string representation is returned"""
+        
+        self.assertTrue(self.user.email, str(self.user))
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_models.py]: @fahadmak  Always add a new line at the end of file

File: authors/apps/authentication/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertEqual(response.data['username'], "jealous")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_edit_user_credentials(self):
+        """Test api can edit user data if user is authenticated"""
+        
+        # authenticate user
+        self.client.force_authenticate(user=self.user)
+
+        user_update = {
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_views.py]: @fahadmak you should create a test fixtures file and put this data there and import it to use here.

**Review by archbishop12 — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/test_base.py (hunk 3 above)
Comment (fahadmak) [reply authors/apps/authentication/tests/test_base.py]: let me work on that

File: authors/apps/authentication/tests/test_base.py (hunk 4 above)
Comment (fahadmak) [reply authors/apps/authentication/tests/test_base.py]: thank you let me work on that

File: authors/apps/authentication/tests/test_views.py (hunk 9 above)
Comment (fahadmak) [reply authors/apps/authentication/tests/test_views.py]: Thanks for  the feedback

**Comment by archbishop12** [comment]
@fahadmak Use Coveralls to see the lines of code you had not tested out. For example the some the edges here are not tested.
screen shot 2019-01-10 at 15 10 12

**Comment by archbishop12** [comment]
@fahadmak Please ensure you checkout the edge cases.

**Comment by archbishop12** [comment]
Please correct your commit message.

File: authors/apps/authentication/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertEqual(response.data['username'], "fahad12")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+
+class UserRetrieveUpdateViewTestCase(BaseTestCase):
+
+    def setUp(self):
+        self.user = User.objects.get(username="fahad12")
+        self.client.force_authenticate(user=self.user)
+        self.url = '/api/user/'
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_views.py]: @fahadmak Please use `reverse` from `django.urls` for this and anywhere else it appears

File: authors/apps/authentication/tests/test_models.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    def setUp(self):
+        """ Creates an instance of the User """
+
+        self.user = User(username="test_user", email="test.sims@andela.com", password="testman101")
+        self.user.save()
+    
+
+    def test_user_creation(self):
+        """test that user is created"""
+        self.assertTrue(self.user.username, "test_user")
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_models.py]: @fahadmak Please reference the data you created in data in data.py eg. `user1_info.user.username` instead of using string `test user` as added here. Do this anyway else you are using data from the ```data.py``` file.

**Review by archbishop12 — APPROVED** [review]

File: authors/apps/authentication/urls.py [hunk 12]
```diff
@@ -5,7 +5,7 @@
 )
 
 urlpatterns = [
-    path('user/', UserRetrieveUpdateAPIView.as_view()),
-    path('users/', RegistrationAPIView.as_view()),
-    path('users/login/', LoginAPIView.as_view()),
+    path('user/', UserRetrieveUpdateAPIView.as_view(), name="detail"),
+    path('users/', RegistrationAPIView.as_view(), name="users"),
```
Comment (archbishop12) [inline authors/apps/authentication/urls.py]: @fahadmak Please change this to more descriptive name
```suggestion
    path('users/', RegistrationAPIView.as_view(), name="register"),
```

Comment (archbishop12) [reply authors/apps/authentication/urls.py]: Let me know when you are done with this so i can merge in this.

**Comment by archbishop12** [comment]
@fahadmak Please run pep8/pycodestyle on the code/files your added to ensure that pep8 standards are met.

**Review by archbishop12 — CHANGES_REQUESTED** [review]
Run pep8 on files added.

File: authors/apps/authentication/urls.py (hunk 12 above)
Comment (fahadmak) [reply authors/apps/authentication/urls.py]: i am done with the changes

**Review by archbishop12 — APPROVED** [review]

**Comment by archbishop12** [comment]
@fahadmak I am merging this.
